### DISCUSSION TITLE
A useful application of `[params]`

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#14627] [Ideas]
An example.
```v
[params]
pub struct Param {
	width  int = 200
	height int = 25
	...
	...
	...
}

pub struct Foo{
mut:
	width  int
	height int
	...
	...
	...
}

pub fn new_foo(param Param) &Foo{
	// If `param` is not given, the value of `param` is set to the default value.
	return &Foo{
		width: param.width
		height: param.height
		...
		...
		...
		...
	}
}

pub fn (mut f Foo) set_params(param Param){
	// If I specify only `height` field, I want to change only the `height` filed value.
	// I want a mechanism to detect that `param` is not set to `width` field.
	param_default := Param{}
	// it would be nice to have a mechanism to simplify this.
        // for check all params and fields.
	if param_default.width != param.width && f.width != param.width{
		f.width = param.width
	}
	if param_default.height != param.height && f.height != param.height{
		f.height = param.height
	}
        ...
        ...
}

fn main(){
	mut f := new_foo() // Nothing is specified -> default value of `Param` is used.
	dump(f)

	mut f2 := new_foo(width: 150)
	dump(f2)
	f2.set_params(height: 10) // If I specify only `height`, I want to change only the `height` value.
	dump(f2)
}
```

it would be useful to use something like `foo.set_params(width: 300, height:30)` as a substitute for setter, since a number of variables can be set at once.
(Personally, though, I think property is more important feature.)

----
https://github.com/vlang/v/blob/master/doc/docs.md#trailing-struct-literal-arguments

**Comment by dumblob** [comment]
Why don't you use sum types?

Anywhere where you want a *mechanism to detect that `param` is not set to...* you should first rewrite it using sum types and see how it looks like. IMHO it'll be quite good.

**Comment by ghost** [comment]
Don't just say it, write concrete code samples.

**Comment by dumblob** [comment]
Ok, in other words the example you gave is IMHO by no means convincing. Could you come up with a (much) better one?

The example you gave is much slower (and verbose!) than blindly assigning `f.width = param.width` without any branching (i.e. without any if-then-else).

**Comment by ghost** [comment]
If you don't give specific sample code, you're just picking a quarrel.

**Comment by dumblob** [comment]
Oh man, it's copy paste! Why to pollute this tracker with such a simple change...

What you wrote is semantically identical **and** faster as this:

```v
[params]
pub struct Param {
	width  int = 200
	height int = 25
	...
	...
	...
}

pub struct Foo{
mut:
	width  int
	height int
	...
	...
	...
}

pub fn new_foo(param Param) &Foo{
	// If `param` is not given, the value of `param` is set to the default value.
	return &Foo{
		width: param.width
		height: param.height
		...
		...
		...
		...
	}
}

pub fn (mut f Foo) set_params(param Param){
	// If I specify only `height` field, I want to change only the `height` filed value.
	// I want a mechanism to detect that `param` is not set to `width` field.
	param_default := Param{}
        // NO NEED TO CHECK ANYTHING
	f.width = param.width
	f.height = param.height
        ...
        ...
}

fn main(){
	mut f := new_foo() // Nothing is specified -> default value of `Param` is used.
	dump(f)

	mut f2 := new_foo(width: 150)
	dump(f2)
	f2.set_params(height: 10) // If I specify only `height`, I want to change only the `height` value.
	dump(f2)
}
```

**Comment by ghost** [comment]
You don't understand what I'm writing at all, and you're giving me your opinion lol
You just impose your ideas on things you don't understand well.

**Comment by Nukiloco** [comment]
There shouldn't be any need to be rude to each other about this 🙂

From looking at both codes, maybe what you are asking for is a set/get feature where you can have a separate setter/getter functions.

For example when changing the height of f2, maybe there can be a feature like this:

```vlang
pub struct Foo{
mut:
	width  int
	@get(get_height) @set(set_height)
	height int
	...
	...
	...
}


pub fn (mut f Foo) get_height(new_height int) int {
	// Run code when getting height
	return height * 2
}

pub fn (mut f Foo) set_height(new_height int) {
	// Run code when setting height
	height = new_height * 4
}

mut f2 := new_foo(width: 150)
f2.height = 10
```

Im not sure if my suggestion goes against vlang's philosophy though many modern languages do allow declaring getters/setters.

**Comment by ghost** [comment]
https://github.com/vlang/v/discussions/12688 For property or getter/setter, my idea is something like this.

This is a little bit different from that, as if I extend [prams], maybe I can do something more useful than property? The idea is that.
In short, the idea is to use a `set_params(p Params)` to set arbitrary multiple property values at once.

Currently, default values are often set as parameters.
And of course, if you pass that structure as an argument, the default value will be used for any fields you don't specify.
However, in GUI, it is often the case that the field has already been assigned another value, so we want to ignore fields that already have a different value from the default, and change the specified field to the specified value.

In short, set_xxx(), set_yyy(), set_zzz() set_...... Think of it as set_params(p Params), which can be done all together.
Currently, [params] struct are given as arguments during initialization, but the idea is that they can also be used to instances.

Currently, I am wondering if this can be done with reflection.
After this post, playXE improved the comptime reflection, so I'll try with that later when I have more time.

In any case, I want property, although some people don't want it.

**Comment by dumblob** [comment]
I'm sorry for my harsh words. I wanted to keep the discussion short and didn't realize it needs more elaboration.

Let me retroactively showcase my train of thoughts.

>However, in GUI, it is often the case that the field has already been assigned another value, so we want to ignore fields that already have a different value from the default, and change the specified field to the specified value.

Yes, this is how I understood your initial textual description. The accompanying example source code:

```v
pub fn (mut f Foo) set_params(param Param){
	// If I specify only `height` field, I want to change only the `height` filed value.
	// I want a mechanism to detect that `param` is not set to `width` field.
	param_default := Param{}
	// it would be nice to have a mechanism to simplify this.
        // for check all params and fields.
	if param_default.width != param.width && f.width != param.width{
		f.width = param.width
	}
	if param_default.height != param.height && f.height != param.height{
		f.height = param.height
	}
        ...
        ...
}
```

didn't demonstrate this though. Instead it seemed to assume that one would **never** call `set_params()` with explicitly using struct's default value as arguments. This sounded so absurd to me that I thought you made a severe mistake in your example and I asked for a better one. Shame on me I didn't specify why I was so perplexed (therefore I mentioned also sum types).

Then when you asked once again for code I thought you simply want to update `f2` and for some reason you don't want to use the [struct update syntax](https://github.com/vlang/v/blob/master/doc/docs.md#struct-update-syntax ):

```v
struct Param{
	height int = 1
	depth int = 3
}
struct Foo{
	height int
	width int
	depth int
}
fn main() {
	tmp := Param{}
	f2 := Foo{ ...tmp, depth: 9, width: 5 }
	println( f2 )
}
```

But why wouldn't you want to use this?

I thought maybe you really didn't want to specify anything at the **call side** (i.e. no keyword arguments) but actually only at the **function definition side**. So I thought your rant was just about cutting down on the hairy branching. I decided to ignore the weird discrepancy of **never** calling `set_params()` with explicitly using struct's default value as arguments and thus understood your example as simply wanting to use a method as an alternative to plain assignment to chosen fields:
```v
fn main() {
	mut f2 := Foo{}
	f2.height = 7
	f2.width = 5
	...
	println( f2 )
}
```

Therefore this misunderstanding.

Let's move on and let me ask you what are the reasons why you don't want to use the *struct update syntax* as demonstrated above?

I can think only of one - namely it's two lines instead of one (due to the `tmp` variable). OTOH I don't think this is a problem in any way and actually it encourages the vertical coding style which seems preferred for *V* code.
